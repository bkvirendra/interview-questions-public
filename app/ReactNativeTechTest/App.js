/**
 * React Native Tech Test
 */

import React from 'react';
import {
  SafeAreaView,
  View,
  StatusBar,
  Text,
  StyleSheet,
} from "react-native";
import { useSelector, useDispatch } from 'react-redux';

import {
  Colors,
} from 'react-native/Libraries/NewAppScreen';

import Header from './components/Header/';
import PetListItem from './components/PetListItem/';
import { getPets } from './selectors'
import PetModal from "./components/PetModal";
import { Actions } from "./reducers";

const App = () => {
  const dispatch = useDispatch();
  const petList = useSelector(getPets);
  const [isModalVisible, setIsModalVisible] = React.useState(false);
  const [activePet, setActivePet] = React.useState(null);

  const onModalHide = () => {
    setIsModalVisible(false);
    setActivePet(null);
  };
  const onModalShow = (pet) => {
    setIsModalVisible(true);
    setActivePet(pet);
  };

  return (
    <SafeAreaView style={styles.background}>
      <StatusBar barStyle={'dark-content'} />
      <Header />
      <View
        style={styles.background}>
        {petList.map(pet => (
          <PetListItem
            onPress={() => onModalShow(pet)}
            pet={pet}
            key={pet.id}
          />
        ))}
        {!petList.length && <Text style={styles.emptyLabel}>Not much to see here!</Text>}
      </View>
      <PetModal
        isVisible={isModalVisible}
        onHide={onModalHide}
        pet={activePet}
        onDelete={() => {
          dispatch({type: Actions.DELETE_PET, payload: activePet});
          onModalHide();
        }}
      />
    </SafeAreaView>
  );
};

export default App;


const styles = StyleSheet.create({
  background: {
    backgroundColor: Colors.lighter,
    flex: 1,
  },
  emptyLabel: {
    margin: 36,
    textAlign: "center",
    fontSize: 21,
    fontWeight: "bold",
  },
});
