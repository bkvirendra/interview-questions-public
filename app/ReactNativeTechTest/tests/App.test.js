/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import * as redux from 'react-redux';

import App from '../App';

describe('Components - App', () => {
  it('renders - light mode', () => {
    const reduxSpy = jest.spyOn(redux, 'useSelector');
    const spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    spyOnUseDispatch.mockReturnValue(null);
    const pets = [
      { id: 0, name: 'Mr. Fluffings', type: 'Dog', feeds: 2 },
      { id: 1, name: 'The Whiskertron', type: 'Cat', feeds: 3 },
      { id: 2, name: 'Dogbert', type: 'Dog', feeds: 3 }
    ];
    reduxSpy.mockReturnValue(pets);
    const tree = renderer
      .create(<App />)
      .toJSON();
    expect(spyOnUseDispatch).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});
