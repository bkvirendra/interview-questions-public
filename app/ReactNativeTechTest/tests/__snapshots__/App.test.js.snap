// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - App renders - light mode 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#F3F3F3",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomWidth": 0.5,
        "marginHorizontal": 12,
        "paddingVertical": 12,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../assets/img/logo_horizontal.png",
        }
      }
      style={
        Object {
          "height": 100.83333333333333,
          "width": 726,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#F3F3F3",
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 92,
          "marginHorizontal": 12,
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "grey",
            "borderRadius": 45,
            "height": 46,
            "marginRight": 16,
            "width": 46,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#444",
              "fontSize": 26,
              "fontWeight": "bold",
            }
          }
        >
          Mr. Fluffings
        </Text>
        <Text
          style={
            Object {
              "color": "#444",
            }
          }
        >
          Dog
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 92,
          "marginHorizontal": 12,
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "grey",
            "borderRadius": 45,
            "height": 46,
            "marginRight": 16,
            "width": 46,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#444",
              "fontSize": 26,
              "fontWeight": "bold",
            }
          }
        >
          The Whiskertron
        </Text>
        <Text
          style={
            Object {
              "color": "#444",
            }
          }
        >
          Cat
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 92,
          "marginHorizontal": 12,
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "grey",
            "borderRadius": 45,
            "height": 46,
            "marginRight": 16,
            "width": 46,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#444",
              "fontSize": 26,
              "fontWeight": "bold",
            }
          }
        >
          Dogbert
        </Text>
        <Text
          style={
            Object {
              "color": "#444",
            }
          }
        >
          Dog
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
